File: jruby/lib/rayz/normal_perturbations.rb
module Rayz
  module NormalPerturbations
    # Sine wave perturbation
    def self.sine_wave(frequency: 10.0, amplitude: 0.1)
      ->(point) {
        Vector.new(
          x: Math.sin(point.y * frequency) * amplitude,
          y: Math.sin(point.z * frequency) * amplitude,
          z: Math.sin(point.x * frequency) * amplitude
        )
      }
    end

    # Quilted pattern perturbation
    def self.quilted(frequency: 5.0, amplitude: 0.15)
      ->(point) {
        u = Math.sin(point.x * frequency)
        v = Math.sin(point.z * frequency)
        magnitude = u * v * amplitude
        Vector.new(x: 0, y: magnitude, z: 0)
      }
    end

    # Simple 3D noise-like perturbation using sine waves
    def self.noise(frequency: 5.0, amplitude: 0.1)
      ->(point) {
        # Use different frequencies for x, y, z to avoid patterns
        nx = Math.sin(point.x * frequency + point.y * frequency * 0.7) * amplitude
        ny = Math.sin(point.y * frequency + point.z * frequency * 0.7) * amplitude
        nz = Math.sin(point.z * frequency + point.x * frequency * 0.7) * amplitude
        Vector.new(x: nx, y: ny, z: nz)
      }
    end

    # Ripple effect
    def self.ripples(center: Point.new(x: 0, y: 0, z: 0), frequency: 10.0, amplitude: 0.1)
      ->(point) {
        # Distance from center
        dx = point.x - center.x
        dz = point.z - center.z
        distance = Math.sqrt(dx * dx + dz * dz)

        # Ripple magnitude based on distance
        magnitude = Math.sin(distance * frequency) * amplitude

        # Perturbation along Y axis
        Vector.new(x: 0, y: magnitude, z: 0)
      }
    end
  end
end

File: python/rayz/normal_perturbations.py
# python/rayz/normal_perturbations.py
from __future__ import annotations

import math
from functools import partial
from typing import Callable

from rayz.tuple import Point, Vector

# These presets return `functools.partial(_module_level_fn, ...)` rather than
# a closure. A closure's inner function isn't picklable (pickle resolves
# functions by qualified module path, and a nested function has none), which
# crashes `Camera.render_parallel`'s `ProcessPoolExecutor` on any material
# using one of these. `partial` bound to a module-level function pickles fine.


def _sine_wave_perturb(frequency: float, amplitude: float, point: Point) -> Vector:
    return Vector(
        math.sin(point.y * frequency) * amplitude,
        math.sin(point.z * frequency) * amplitude,
        math.sin(point.x * frequency) * amplitude,
    )


def sine_wave(frequency: float = 10.0, amplitude: float = 0.1) -> Callable[[Point], Vector]:
    return partial(_sine_wave_perturb, frequency, amplitude)


def _quilted_perturb(frequency: float, amplitude: float, point: Point) -> Vector:
    u = math.sin(point.x * frequency)
    v = math.sin(point.z * frequency)
    magnitude = u * v * amplitude
    return Vector(0, magnitude, 0)


def quilted(frequency: float = 5.0, amplitude: float = 0.15) -> Callable[[Point], Vector]:
    return partial(_quilted_perturb, frequency, amplitude)


def _noise_perturb(frequency: float, amplitude: float, point: Point) -> Vector:
    nx = math.sin(point.x * frequency + point.y * frequency * 0.7) * amplitude
    ny = math.sin(point.y * frequency + point.z * frequency * 0.7) * amplitude
    nz = math.sin(point.z * frequency + point.x * frequency * 0.7) * amplitude
    return Vector(nx, ny, nz)


def noise(frequency: float = 5.0, amplitude: float = 0.1) -> Callable[[Point], Vector]:
    return partial(_noise_perturb, frequency, amplitude)


def _ripples_perturb(cx: float, cz: float, frequency: float, amplitude: float, point: Point) -> Vector:
    dx = point.x - cx
    dz = point.z - cz
    distance = math.sqrt(dx * dx + dz * dz)
    magnitude = math.sin(distance * frequency) * amplitude
    return Vector(0, magnitude, 0)


def ripples(
    center: Point | None = None,
    frequency: float = 10.0,
    amplitude: float = 0.1,
) -> Callable[[Point], Vector]:
    if center is None:
        center = Point(0, 0, 0)
    return partial(_ripples_perturb, center.x, center.z, frequency, amplitude)
